[Lambda] Lambda Cloud SkyPilot provisioner

kmushegi opened the pull request:
This PR implements the SkyPilot provisioner for Lambda Cloud.



Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
    - [x]  Ran multiple tests against Lambda Cloud.
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+        instances[instance_id] = [
+            common.InstanceInfo(
+                instance_id=instance_id,
+                # TODO (kmushegi): check if this is correct;
+                # external ip is preferred
+                internal_ip='',
```
Could we also check if is there a private IP for Lambda Cloud? From the previous implementation seems like they have:

https://github.com/skypilot-org/skypilot/blob/c7ee6a22c2fbb912bbe8f83b2931062e6b4e4490/sky/skylet/providers/lambda_cloud/node_provider.py#L232-L247

kmushegi replied:
I removed this somewhat purposefully since this check to break intermittently. it relies on the `docker0` network interface to be up and on some instances it is not up.

would it make sense to make a best effort to grab the internal ip? lmk

cblmemo replied:
I think the private ip would be preferred by some user who concerned about privacy; cc @Michaelvll for inputs here

Michaelvll replied:
Our previous implementation for getting private IP is indeed quite hacky. Just wondering if Lambda has updated their API to include private IP? It seems to me that they do have it in their doc: https://cloud.lambdalabs.com/api/v1/docs#operation/listInstances

Maybe we should use that to see if we can connect our underlying ray cluster through that private IP?

kmushegi replied:
Ohhh word, great!

kmushegi replied:
this worked and fixed the multi node startup issues as well :)

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+                internal_ip='',
+                external_ip=instance_info['ip'],
+                ssh_port=22,
+                tags={},
```
Why not include tag information here? I suppose there are tags attached to the Lambda VM?

https://github.com/skypilot-org/skypilot/blob/c7ee6a22c2fbb912bbe8f83b2931062e6b4e4490/sky/skylet/providers/lambda_cloud/node_provider.py#L101-L122

kmushegi replied:
Lambda's API does not support tagging afaik so these tags would be internal to Skypilot? happy to keep them in but looking at the runpod implementation, tags are left empty: https://github.com/skypilot-org/skypilot/blob/master/sky/provision/runpod/instance.py#L173

cblmemo replied:
Yeah that makes sense. LGTM

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    # Do nothing because we assume the user has updated their firewall
+    # rules to allow these ports
```
What does this comment mean? IIUC we does not support opening ports on Lambda?

kmushegi replied:
that's correct, leftover comment from an initial internal implementation of this. removing and raising not implemented.

cblmemo commented on sky/provision/lambda_cloud/lambda_utils.py:
```diff
@@ -163,27 +166,27 @@ def create_instances(self,
-                instance_ids[0]  # TODO(ewzeng) don't hardcode
-            ]
-        })
+        data = json.dumps({'instance_ids': [instance_ids[0]]
+                          }  # TODO(ewzeng) don't hardcode
+                         )
```
```suggestion
        # TODO(ewzeng) don't hardcode
        data = json.dumps({'instance_ids': [instance_ids[0]]})
```
nit

kmushegi replied:
good catch here esp since we're always passing down a list now so need to grab all instance ids

cblmemo commented on sky/provision/lambda_cloud/lambda_utils.py:
```diff
@@ -131,20 +132,22 @@ def __init__(self) -> None:
         # See https://docs.lambdalabs.com/cloud/rate-limiting/ for more.
-        available_regions = self.list_catalog()[instance_type]\
-                ['regions_with_capacity_available']
+        available_regions = self.list_catalog(
+        )[instance_type]['regions_with_capacity_available']
```
```suggestion
        available_regions = (self.list_catalog(
        )[instance_type]['regions_with_capacity_available'])
```
nit: and re-run formatter

kmushegi replied:
done

cblmemo reviewed (commented):
Thanks for this amazing PR @kmushegi ! 🚀 It would be really useful to move Lambda to the new provisioner and speed up provisioning a lot. Left some comments to discuss!

romilbhardwaj commented:
Thanks @kmushegi! 

Trying this out, ran into this error when trying to launch

```
$ sky launch -c lamb --num-nodes 2 --cloud lambda -- echo hi
...
I 08-27 16:16:02 provisioner.py:65] Launching on Lambda us-east-1 (all zones)
E 08-27 16:16:02 provisioner.py:80] Failed to configure 'lamb' on Lambda Region(name='us-east-1') (all zones) with the following error:
E 08-27 16:16:02 provisioner.py:80] AssertionError: Unknown provider: lambda
D 08-27 16:16:02 provisioner.py:171] Failed to provision 'lamb' on Lambda (all zones).
D 08-27 16:16:02 provisioner.py:173] bulk_provision for 'lamb' failed. Stacktrace:
D 08-27 16:16:02 provisioner.py:173] Traceback (most recent call last):
D 08-27 16:16:02 provisioner.py:173]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/provisioner.py", line 165, in bulk_provision
D 08-27 16:16:02 provisioner.py:173]     return _bulk_provision(cloud, region, zones, cluster_name,
D 08-27 16:16:02 provisioner.py:173]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/provisioner.py", line 76, in _bulk_provision
D 08-27 16:16:02 provisioner.py:173]     config = provision.bootstrap_instances(provider_name, region_name,
D 08-27 16:16:02 provisioner.py:173]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/__init__.py", line 44, in _wrapper
D 08-27 16:16:02 provisioner.py:173]     assert module is not None, f'Unknown provider: {module_name}'
D 08-27 16:16:02 provisioner.py:173] AssertionError: Unknown provider: lambda
```

kmushegi commented:
thanks for the reviews folks, will try to address asap

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+from sky import sky_logging
+from sky import status_lib
+from sky.provision import common
+from sky.provision.lambda_cloud.lambda_utils import LambdaCloudClient
+from sky.provision.lambda_cloud.lambda_utils import LambdaCloudError
```
done, thanks for pointing this out!

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+
+    filtered_instances = {}
+    for instance in instances:
+        if status_filters is not None and instance[
+                'status'] not in status_filters:
```
done.

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+
+def _get_ssh_key_name(prefix: str = '') -> str:
+    lambda_client = _get_lambda_client()
+    public_key_path = os.path.expanduser(auth.PUBLIC_SSH_KEY_PATH)
+    with open(public_key_path, 'r', encoding='utf-8') as f:
+        public_key = f.read()
```
makes sense!

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+                instance_type=config.node_config['InstanceType'],
+                region=region,
+                name=f'{cluster_name_on_cloud}-{node_type}',
+                quantity=1,
```
yeah for sure, I was gonna leave it for next iteration but might as well. done

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+                name=f'{cluster_name_on_cloud}-{node_type}',
+                quantity=1,
+                ssh_key_name=ssh_key_name,
+            )[0]
```
create_instances returns a `List[str]` of instance_ids so there's nothing to unpack?

cblmemo replied:
Oh i see. If that is the case, it would be great to assert the returned list's length is 1 ;)

kmushegi replied:
Good point to check that, is raising valuerror compliant enough with the coding style? or assert?

cblmemo replied:
For this case, it is really unlikely that the returned list has a length that is not 1, as we specify `quantity=1` here - I think assertion would fits the semantic better?

kmushegi replied:
makes sense, done

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+        created_instance_ids.append(instance_id)
+        if head_instance_id is None:
+            head_instance_id = instance_id
+        time.sleep(10)  # Avoid api rate limits
```
this happens now cause we're creating instances in bulk :)

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+            head_instance_id = instance_id
+        time.sleep(10)  # Avoid api rate limits
+
+    while True:
+        instances = _filter_instances(cluster_name_on_cloud, ['active'])
+        ready_instance_cnt = 0
+        for instance_id, instance in instances.items():
+            if instance.get('status') == 'active':
+                ready_instance_cnt += 1
+        if ready_instance_cnt == config.count:
+            break
+
+        time.sleep(POLL_INTERVAL)
```
yes 😄  good catch!

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+from sky.utils import common_utils
+from sky.utils import ux_utils
+
+POLL_INTERVAL = 5
```
setting to 1 makes sense as retry with backoff initially delays by 10 seconds already, no need to add more :)

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+
+        time.sleep(POLL_INTERVAL)
+
+    assert head_instance_id is not None, 'head_instance_id should not be None'
```
the loop is not gone but adding the assert right after where the head instance id should be available.

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+        region=region,
+        zone=None,
+        head_instance_id=head_instance_id,
+        resumed_instance_ids=[],
```
hmm what does a "resumed instance" mean? since lambda does not support stopping, I figured this would always be empty?

cblmemo replied:
Oh that makes sense. LGTM!

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,271 @@
+    del provider_config
+    lambda_client = _get_lambda_client()
+    instances = _filter_instances(cluster_name_on_cloud, None)
+    # TODO (kmushegi): terminate all instances together since
+    # remove_instances accepts multiple instance ids
```
ya might as well! :)

kmushegi commented:
> Thanks @kmushegi!
> 
> Trying this out, ran into this error when trying to launch
> 
> ```
> $ sky launch -c lamb --num-nodes 2 --cloud lambda -- echo hi
> ...
> I 08-27 16:16:02 provisioner.py:65] Launching on Lambda us-east-1 (all zones)
> E 08-27 16:16:02 provisioner.py:80] Failed to configure 'lamb' on Lambda Region(name='us-east-1') (all zones) with the following error:
> E 08-27 16:16:02 provisioner.py:80] AssertionError: Unknown provider: lambda
> D 08-27 16:16:02 provisioner.py:171] Failed to provision 'lamb' on Lambda (all zones).
> D 08-27 16:16:02 provisioner.py:173] bulk_provision for 'lamb' failed. Stacktrace:
> D 08-27 16:16:02 provisioner.py:173] Traceback (most recent call last):
> D 08-27 16:16:02 provisioner.py:173]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/provisioner.py", line 165, in bulk_provision
> D 08-27 16:16:02 provisioner.py:173]     return _bulk_provision(cloud, region, zones, cluster_name,
> D 08-27 16:16:02 provisioner.py:173]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/provisioner.py", line 76, in _bulk_provision
> D 08-27 16:16:02 provisioner.py:173]     config = provision.bootstrap_instances(provider_name, region_name,
> D 08-27 16:16:02 provisioner.py:173]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/__init__.py", line 44, in _wrapper
> D 08-27 16:16:02 provisioner.py:173]     assert module is not None, f'Unknown provider: {module_name}'
> D 08-27 16:16:02 provisioner.py:173] AssertionError: Unknown provider: lambda
> ```

Fixed this, missed a change to commit initially.

Moving onto some testing

cblmemo commented:
Hi @kmushegi I'm trying this today and encountered the following error. What does `quantity: Input should be less than or equal to 1` mean here? Can we have a more informative error message here?

```bash
sky launch --cloud lambda --num-nodes 3 -c lmd-3node
I 09-11 14:16:56 optimizer.py:719] == Optimizer ==
I 09-11 14:16:56 optimizer.py:730] Target: minimizing cost
I 09-11 14:16:56 optimizer.py:742] Estimated cost: $2.2 / hour
I 09-11 14:16:56 optimizer.py:742] 
I 09-11 14:16:56 optimizer.py:867] Considered resources (3 nodes):
I 09-11 14:16:56 optimizer.py:937] ------------------------------------------------------------------------------------------
I 09-11 14:16:56 optimizer.py:937]  CLOUD    INSTANCE     vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 09-11 14:16:56 optimizer.py:937] ------------------------------------------------------------------------------------------
I 09-11 14:16:56 optimizer.py:937]  Lambda   gpu_1x_a10   30      200       A10:1          us-east-1     2.25          ✔     
I 09-11 14:16:56 optimizer.py:937] ------------------------------------------------------------------------------------------
I 09-11 14:16:56 optimizer.py:937] 
Launching a new cluster 'lmd-3node'. Proceed? [Y/n]: 
I 09-11 14:16:57 cloud_vm_ray_backend.py:4397] Creating a new cluster: 'lmd-3node' [3x Lambda(gpu_1x_a10, {'A10': 1})].
I 09-11 14:16:57 cloud_vm_ray_backend.py:4397] Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
I 09-11 14:16:57 cloud_vm_ray_backend.py:1314] To view detailed progress: tail -n100 -f /home/memory/sky_logs/sky-2024-09-11-14-16-56-536636/provision.log
I 09-11 14:16:58 provisioner.py:65] Launching on Lambda us-east-1 (all zones)
W 09-11 14:17:02 instance.py:117] run_instances error: global/invalid-parameters: quantity: Input should be less than or equal to 1
W 09-11 14:17:05 cloud_vm_ray_backend.py:2003] sky.exceptions.ResourcesUnavailableError: Failed to acquire resources in all zones in us-east-1. Try changing resource requirements or use another region.
W 09-11 14:17:05 cloud_vm_ray_backend.py:2012] 
W 09-11 14:17:05 cloud_vm_ray_backend.py:2012] Provision failed for 3x Lambda(gpu_1x_a10, {'A10': 1}) in us-east-1. Trying other locations (if any).
```

cblmemo reviewed (commented):
Thanks for the fix @kmushegi ! I tested this PR and all of launch/terminate/multinode works well. Left some nits and after that it should be ready to go!

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,277 @@
+        'booting': status_lib.ClusterStatus.INIT,
+        'active': status_lib.ClusterStatus.UP,
+        'unhealthy': status_lib.ClusterStatus.INIT,
+        'terminating': status_lib.ClusterStatus.STOPPED,
+        'terminated': status_lib.ClusterStatus.STOPPED,
```
yeah good call on not even showing it, removing terminated and mapping terminating to init.

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,277 @@
+
+    if head_instance_id is None:
+        instance_ids = launch_nodes('head', 1)
+        if len(instance_ids) != 1:
+            raise RuntimeError(
+                f'Expected exactly one instance, got {len(instance_ids)}')
```
sg.

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,277 @@
+    created_instance_ids = []
+    ssh_key_name = _get_ssh_key_name()
+
+    def launch_nodes(node_type: str, quantity: int):
```
done

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,277 @@
+        provider_config=provider_config,
+        custom_ray_options={
+            'use_external_ip': True,
+        },
```
was vestigial from dev pre internal ips being added. removed

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,277 @@
+
+
+def _filter_instances(cluster_name_on_cloud: str,
+                      status_filters: Optional[List[str]]) -> Dict[str, Any]:
```
done

kmushegi commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,277 @@
+    try:
+        logger.debug(
+            f'Terminating instances {", ".join(instance_ids_to_terminate)}')
+        lambda_client.remove_instances(*instance_ids_to_terminate)
```
done.

kmushegi commented on sky/provision/lambda_cloud/lambda_utils.py:
```diff
@@ -163,27 +166,25 @@ def create_instances(self,
+        )
         return response.json().get('data', []).get('instance_ids', [])
 
     def remove_instances(self, *instance_ids: str) -> Dict[str, Any]:
```
done.

kmushegi commented:
> Hi @kmushegi , just checking the status of this PR - Is this still in your schedule? If not we can take it from here :))

apologies for the delay had a bunch of things come up, just addressed all reviews and waiting for CI pass. Upon CI pass should be good for final testing.

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,274 @@
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    raise NotImplementedError()
```
add some description as stop_instances?

kmushegi replied:
good call, done

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,274 @@
+
+    if head_instance_id is None:
+        instance_ids = launch_nodes('head', 1)
+        assert len(
+            instance_ids
+        ) == 1, f'Expected exactly one instance, got {len(instance_ids)}'
```
```suggestion
        assert len(instance_ids) == 1
```
we probably can omit the message as this is very unlikely to happen ;) (e.g. API from lambda lab failed)

kmushegi replied:
yeah sgtm

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,274 @@
+
+    assert head_instance_id is not None, 'head_instance_id should not be None'
+
+    worker_node_count = to_start_count - 1
```
```suggestion
    worker_node_count = to_start_count - 1
```
this should only minus one if the head instance id is none?

kmushegi replied:
hmm we assert above that head instance id is not none?

cblmemo replied:
Sorry for the confusion. I mean this

<img width="870" alt="image" src="https://github.com/user-attachments/assets/77f068f3-ed22-4a4d-a07c-582a871b8c4c">

logically it is possible that our head instance is provisioned and this time we only wants to add some workers. in this case, the `to_start_count` should be equal to `worker_node_count`. Though practically this wont appear in our system but we should do it for future expansion

kmushegi replied:
so are you suggesting smth like 

```python
if head_instance_id is None:
    worker_node_count = to_start_count - 1
else:
    worker_node_count = to_start_count
```

Michaelvll replied:
See this comment: https://github.com/skypilot-org/skypilot/pull/3865#discussion_r1798724438

cblmemo commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,274 @@
+        'terminating': status_lib.ClusterStatus.INIT,
+    }
+    statuses: Dict[str, Optional[status_lib.ClusterStatus]] = {}
+    for instance_id, instance in instances.items():
+        status = status_map.get(instance['status'])
+        if non_terminated_only and status is None:
+            continue
+        statuses[instance_id] = status
```
Should we filter the active status if `non_terminated_only is True`?

kmushegi replied:
not sure I understand the question

cblmemo replied:
oh nvm. I misread the argument name. looks great!

cblmemo reviewed (commented):
Thanks for the prompt fix @kmushegi ! I tested launch/execution/termination and all works smoothly. It mostly looks good to me! cc @Michaelvll for a final check here

Michaelvll commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,272 @@
+            f'Cluster {cluster_name_on_cloud} already has '
+            f'{len(exist_instances)} nodes, but {config.count} are required.')
+    if to_start_count == 0:
+        if head_instance_id is None:
+            raise RuntimeError(
+                f'Cluster {cluster_name_on_cloud} has no head node.')
```
nit: Instead of error out, can we just patch one of the nodes to make it head?

kmushegi replied:
I don't have bandwidth this week to implement this, if someone wants to take a stab at it pls do cause I like the idea

cblmemo replied:
@Michaelvll I'm trying to give this a stab but just realized that there might be different runtime on worker node and head node (e.g. ray config) which will involve complex pattern recognition on the node runtime. Also, the only scenario i can think of that will trigger this case is user manually change the instance name on the console, which is a rare case, and seems like runpod's implementation takes the similar approach of directly raising it. Do you think we could leave this to another PR?

https://github.com/skypilot-org/skypilot/blob/a4e2fcd438d70373377c85bcbec1b185ef04c99f/sky/provision/runpod/instance.py#L68-L70

Just filed an issue for that: #4087

Michaelvll commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,272 @@
+            logger.warning(f'run_instances error: {e}')
+            raise
+
+    if head_instance_id is None:
+        instance_ids = launch_nodes('head', 1)
+        assert len(instance_ids) == 1
+        created_instance_ids.append(instance_ids[0])
+        head_instance_id = instance_ids[0]
+
+    assert head_instance_id is not None, 'head_instance_id should not be None'
+
+    worker_node_count = to_start_count - 1
```
Seems like the logic should be the following
```suggestion
    if head_instance_id is None:
        instance_ids = launch_nodes('head', 1)
        assert len(instance_ids) == 1
        created_instance_ids.append(instance_ids[0])
        head_instance_id = instance_ids[0]
        to_start_count -= 1

    assert head_instance_id is not None, 'head_instance_id should not be None'
    
    worker_node_count = to_start_count
```

Michaelvll commented on sky/provision/lambda_cloud/instance.py:
```diff
@@ -0,0 +1,272 @@
+    del cluster_name_on_cloud, ports, provider_config  # Unused.
+
+
+def query_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    head_ip: Optional[str] = None,
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> Dict[int, List[common.Endpoint]]:
+    """See sky/provision/__init__.py"""
+    del cluster_name_on_cloud, provider_config  # Unused.
+    return common.query_ports_passthrough(ports, head_ip)
```
This is not required, since `query_ports` has default implementation. We can remove `query_ports` for lambda entirely.

kmushegi replied:
done.

kmushegi replied:
actually mypy ci fails w/o this, putting it back

cblmemo replied:
I think remove this line would resolve the problem:

<img width="1314" alt="image" src="https://github.com/user-attachments/assets/545782c2-3c92-4595-a3f3-d079d7172f4d">

kmushegi replied:
sure just pushed

Michaelvll reviewed (commented):
Thanks again for adding this @kmushegi!

cblmemo reviewed (commented):
Thanks for the awesome work @kmushegi ! I tested it and it all runs like smooth. There are only one minor issue: for lambda cloud cluster that launched before this PR, running `sky status --refresh` makes it in `INIT` state. Besides that it should be ready to go! Could you help take a look at this?

```bash
(sky) ➜  skypilot git:(master) sky launch --cloud lambda -c lmd-old --gpus A100 nvidia-smi
Task from command: nvidia-smi
Considered resources (1 node):
------------------------------------------------------------------------------------------------
 CLOUD    INSTANCE           vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
------------------------------------------------------------------------------------------------
 Lambda   gpu_1x_a100_sxm4   30      200       A100:1         us-east-1     1.29          ✔     
------------------------------------------------------------------------------------------------
Multiple Lambda instances satisfy A100:1. The cheapest Lambda(gpu_1x_a100_sxm4, {'A100': 1}) is considered among:
['gpu_1x_a100_sxm4', 'gpu_1x_a100'].
To list more details, run: sky show-gpus A100

Launching a new cluster 'lmd-old'. Proceed? [Y/n]: 
⚙︎ Launching on Lambda us-east-1.
  Head VM is up.
✓ Cluster launched: 'lmd-old'.  View logs at: ~/sky_logs/sky-2024-10-16-11-59-04-228003/provision.log
⚙︎ Job submitted, ID: 1
├── Waiting for task resources on 1 node.
└── Job started. Streaming logs... (Ctrl-C to exit log streaming; job will not be killed)
(sky-cmd, pid=4130) Wed Oct 16 19:04:24 2024       
(sky-cmd, pid=4130) +---------------------------------------------------------------------------------------+
(sky-cmd, pid=4130) | NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
(sky-cmd, pid=4130) |-----------------------------------------+----------------------+----------------------+
(sky-cmd, pid=4130) | GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
(sky-cmd, pid=4130) | Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
(sky-cmd, pid=4130) |                                         |                      |               MIG M. |
(sky-cmd, pid=4130) |=========================================+======================+======================|
(sky-cmd, pid=4130) |   0  NVIDIA A100-SXM4-40GB          On  | 00000000:07:00.0 Off |                    0 |
(sky-cmd, pid=4130) | N/A   32C    P0              44W / 400W |      4MiB / 40960MiB |      0%      Default |
(sky-cmd, pid=4130) |                                         |                      |             Disabled |
(sky-cmd, pid=4130) +-----------------------------------------+----------------------+----------------------+
(sky-cmd, pid=4130)                                                                                          
(sky-cmd, pid=4130) +---------------------------------------------------------------------------------------+
(sky-cmd, pid=4130) | Processes:                                                                            |
(sky-cmd, pid=4130) |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
(sky-cmd, pid=4130) |        ID   ID                                                             Usage      |
(sky-cmd, pid=4130) |=======================================================================================|
(sky-cmd, pid=4130) |  No running processes found                                                           |
(sky-cmd, pid=4130) +---------------------------------------------------------------------------------------+
✓ Job finished (status: SUCCEEDED).
Shared connection to 150.136.118.84 closed.

📋 Useful Commands
Job ID: 1
├── To cancel the job:          sky cancel lmd-old 1
├── To stream job logs:         sky logs lmd-old 1
└── To view job queue:          sky queue lmd-old

Cluster name: lmd-old
├── To log into the head VM:    ssh lmd-old
├── To submit a job:            sky exec lmd-old yaml_file
├── To stop the cluster:        sky stop lmd-old
└── To teardown the cluster:    sky down lmd-old

(sky) ➜  skypilot git:(master) gsw feat/oss-lambda-cloud-new-provisioner
Switched to branch 'feat/oss-lambda-cloud-new-provisioner'
(sky) ➜  skypilot git:(feat/oss-lambda-cloud-new-provisioner) sst -r                                                     
Clusters
Refreshing status for 2 clusters ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--instance_id: 25d1811e8e004be7975979a6b3dd4923, status: ClusterStatus.UP
NAME                           LAUNCHED     RESOURCES                                                 STATUS   AUTOSTOP  COMMAND                          
lmd-old                        36 secs ago  1x Lambda(gpu_1x_a100_sxm4, {'A100': 1})                  INIT     -         sky launch --cloud lambda -c...  
sky-serve-controller-402b1bba  1 hr ago     1x AWS(m6i.xlarge, disk_size=200, ports=['30001-30020'])  STOPPED  10m       sky serve up examples/ser...     

Managed jobs
No in-progress managed jobs. (See: sky jobs -h)

Services
No live services.
```

cblmemo reviewed (approved):
Thanks for contributing this awesome work @kmushegi ! I filed an issue to keep track of the issue above. Merging it now to unlock it first!
